### PR TITLE
DB-6591: implement check_table system procedures(2.5)

### PR DESCRIPTION
**Pull request by jyuanca**
Description is available at https://splicemachine.atlassian.net/browse/DB-6591

### CODE REVIEW

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                messages.add(indexKeyTemplate.getClone().toString() + "=>" + tuple._1);
+                count++;
+            }
+        }
+        if (count >= maxCheckTableError) {
+            count = duplicateIndexRdd.count();
+            messages.add("...");
+        }
+
+        messages.add(0, String.format("The following %d indexes are duplicates:", count));
```
Comment (snowspeeder) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java]: you mean to say indexes "rows" are duplicate?

File: splice_machine/src/main/java/com/splicemachine/derby/stream/control/ControlTableChecker.java [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+            if (!tableData.containsKey(baseRowId)) {
+                List<byte[]> rows = indexData.get(baseRowId);
+                result.putAll(baseRowId, rows);
+                invalidIndexCount += rows.size();
+            }
+        }
+
+        int i = 0;
+        if (invalidIndexCount > 0) {
+            messages.add(String.format("The following %d indexes are invalid:", invalidIndexCount));
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/control/ControlTableChecker.java]: you mean to say indexes "rows" are invalid?

File: splice_machine/src/main/java/com/splicemachine/derby/stream/control/ControlTableChecker.java [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        for (String baseRowId : indexData.keySet()) {
+            List<byte[]> rows = indexData.get(baseRowId);
+            if (rows.size() > 1) {
+                duplicateIndexCount += rows.size();
+                result.putAll(baseRowId, rows);
+            }
+        }
+        int i = 0;
+        if (duplicateIndexCount > 0) {
+            messages.add(String.format("The following %d indexes are duplicates:", duplicateIndexCount));
```
Comment (snowspeeder) [inline splice_machine/src/main/java/com/splicemachine/derby/stream/control/ControlTableChecker.java]: you mean to say indexes "rows" are duplicate?

File: hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    /**
+     * Check base table whether all rows are indexed
+     * @param table
+     * @param index
+     * @return
+     * @throws StandardException
+     */
+    private List<String> checkMissingIndexes(PairDataSet table, PairDataSet index) throws StandardException {
+        List<String> messages = Lists.newLinkedList();
+        SpliceSpark.pushScope(String.format("Check unindexed rows from table %s.%s", schemaName, tableName));
```
Comment (yxia92) [inline hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/SparkTableChecker.java]: A call to popScope() is missing in this function.

File: splice_machine/src/main/java/com/splicemachine/derby/impl/storage/CheckTableJob.java [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+                        .transaction(txn)
+                        .scan(createScan(txn))
+                        .tableVersion(tableVersion)
+                        .template(templateRow)
+                        .keyColumnEncodingOrder(columnOrdering)
+                        .keyColumnTypes(ScanOperation.getKeyFormatIds(columnOrdering, formatIds))
+                        .accessedKeyColumns(accessedKeyColumns)
+                        .keyDecodingMap(ScanOperation.getKeyDecodingMap(accessedKeyColumns, columnOrdering, baseColumnMap))
+                        .rowDecodingMap(ScanOperation.getRowDecodingMap(accessedKeyColumns, columnOrdering, baseColumnMap))
+                        .baseColumnMap(baseColumnMap)
```
[the remark this answers is not in the document]
Comment (yxia92) [reply splice_machine/src/main/java/com/splicemachine/derby/impl/storage/CheckTableJob.java]: On a second thought, defaultRow is not needed, as we are only returning the primary key or rowid from each row, and we won't be touching the columns with non-populated values. 
@jyuanca  you may ignore the above comment.

**Comment by yxia92** [comment]
@jyuanca This tool may not work with the index excluding default/null feature where the index rows are expected to be smaller than the number of base table rows.

**Pull request closed without merging**